File: benchmarks/clusters/idris.jean_zay/plot_results.py
#!/usr/bin/env python3
# encoding: utf-8

import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt

from typing import NamedTuple

MPI_TAG = "intel_19.0.4"


class Benchmark(NamedTuple):
    name: str
    filename_fmt: str
    plot_style: dict = dict()


BENCH_PENCILS = Benchmark('PencilFFTs', 'results/PencilFFTs_N{N}_{MPI}.dat',
                          plot_style=dict(color='C0'))

BENCH_P3DFFT = Benchmark('P3DFFT', 'results/P3DFFT3_N{N}_{MPI}.dat',
                         plot_style=dict(color='C1'))


RESOLUTIONS = (512, 1024)

PARAMS_PI = 0  # with dimension permutations + Isend/Irecv
PARAMS_PA = 1
PARAMS_NI = 2
PARAMS_NA = 3  # no dimension permutations + Alltoallv
PARAMS_DEFAULT = PARAMS_PI

PARAMS_ALL = (
    PARAMS_PI,
    PARAMS_PA,
    PARAMS_NI,
    PARAMS_NA,
)

STYLE_PERM = dict(color='C0')
STYLE_NOPERM = dict(color='C3')
STYLE_ISEND = dict(ls='-')
STYLE_ALLTO = dict(ls='--')

STYLE_PARAMS = (
    dict(**STYLE_PERM, **STYLE_ISEND),
    dict(**STYLE_PERM, **STYLE_ALLTO),
    dict(**STYLE_NOPERM, **STYLE_ISEND),
    dict(**STYLE_NOPERM, **STYLE_ALLTO),
)

STYLE_RESOLUTION = {
    512: dict(marker='o'),
    1024: dict(marker='^')
}

STYLE_IDEAL = dict(color='black', ls=':', label='ideal')


def plot_from_file(ax: plt.Axes, bench: Benchmark, resolution: int,
                   speedup=False, params=PARAMS_ALL, plot_kw=None):
    ax.set_xscale('log', basex=2)
    ax.set_yscale('log', basey=(2 if speedup else 10))

    is_pencilffts = bench is BENCH_PENCILS
    filename = bench.filename_fmt.format(N=resolution, MPI=MPI_TAG)
    data = np.loadtxt(filename)
    Nxyz = data[:, 0:3]
    N = Nxyz[0, 0]
    assert np.all(Nxyz == N)  # they're all the same
    procs = data[:, 3]
    # proc_dims = data[:, 4:6]
    times = data[:, 7:]
    Ntimes = times.shape[1]

    times_ideal = np.outer(1.0 / procs[:], times[0, :]) * procs[0]

    if speedup:
        times[:, :] = times[0, :] / times[:, :]
        times_ideal[:, :] = times_ideal[0, :] / times_ideal[:, :]

    for n in range(Ntimes):
        kw = bench.plot_style.copy()
        if is_pencilffts:
            if n not in params:
                continue
            if len(params) > 1:
                kw.update(STYLE_PARAMS[n])
        if plot_kw is not None:
            kw.update(plot_kw)
        ax.plot(procs, times[:, n], **kw)

    if is_pencilffts:
        ax.plot(procs, times_ideal[:, 0], **STYLE_IDEAL)


def compare_params(ax: plt.Axes, resolution, bench=BENCH_PENCILS,
                   params=PARAMS_ALL, plot_kw=None):
    filename = bench.filename_fmt.format(N=resolution, MPI=MPI_TAG)
    data = np.loadtxt(filename)
    Nxyz = data[:, 0:3]
    N = Nxyz[0, 0]
    assert np.all(Nxyz == N)  # they're all the same
    procs = data[:, 3]
    # proc_dims = data[:, 4:6]
    times = data[:, 7:]
    Ntimes = times.shape[1]

    # Compare with the first column (default params)
    times /= times[:, 0:1]

    for n in params:
        kw = STYLE_PARAMS[n].copy()
        if plot_kw is not None:
            kw.update(plot_kw)
        ax.plot(procs, times[:, n], **kw)


def add_legend(ax: plt.Axes, styles, labels, **kwargs):
    leg = ax.get_legend()
    lines = [mpl.lines.Line2D([], [], **style) for style in styles]
    ax.legend(lines, labels, **kwargs)
    if leg is not None:
        ax.add_artist(leg)


def legend_benchmarks(ax: plt.Axes, bench):
    styles = [*map(lambda b: b.plot_style, bench), STYLE_IDEAL]
    labels = [*map(lambda b: b.name, bench), 'Ideal']
    add_legend(ax, styles, labels, loc='upper right')


def legend_params(ax: plt.Axes, params):
    styles = map(lambda p: STYLE_PARAMS[p], params)


def legend_resolution(ax: plt.Axes, resolutions, loc='lower left'):
    styles = [dict(color='black', ls='-', **STYLE_RESOLUTION[n])
              for n in resolutions]
    labels = map(lambda N: f'{N}³', resolutions)
    add_legend(ax, styles, labels, loc=loc, title='Resolution')


def plot_comparison(ax: plt.Axes, N, legend=False, **kwargs):
    st = STYLE_RESOLUTION[N]
    bench = (BENCH_PENCILS, BENCH_P3DFFT)
    for b in bench:
        plot_from_file(ax, b, N, params=(PARAMS_DEFAULT, ),
                       plot_kw=st, **kwargs)
    if legend:
        legend_benchmarks(ax, bench)


def plot_params(ax: plt.Axes, N, legend=False, **kwargs):
    ax.set_xscale('log', basex=2)
    st = STYLE_RESOLUTION[N]
    params = (PARAMS_PI, PARAMS_PA)
    compare_params(ax, N, BENCH_PENCILS, params, plot_kw=st)


def plot_timings(ax: plt.Axes, resolutions, plot_func=plot_comparison,
                 speedup=False):
    for n, N in enumerate(resolutions):
        plot_func(ax, N, speedup=speedup, legend=(n == 0))

    if speedup:
        ylab = 'Speedup'
        loc = 'upper left'
    else:
        ylab = 'Time (milliseconds)'
        loc = 'lower left'

    legend_resolution(ax, resolutions, loc)
    ax.set_xlabel('MPI processes')

    # Show '1024' instead of '2^10'
    for axis in (ax.xaxis, ax.yaxis):
        axis.set_major_formatter(mpl.ticker.ScalarFormatter())

    ax.set_ylabel(ylab)


fig, ax = plt.subplots()
plot_timings(ax, RESOLUTIONS, plot_comparison)
fig.savefig('timing_comparison.svg')

fig, ax = plt.subplots()
plot_timings(ax, RESOLUTIONS, plot_params)
# fig.savefig('timing_params.svg')

plt.show()

File: benchmarks/clusters/idris.jean_zay/plot_results.jl
#!/usr/bin/env julia

using DelimitedFiles: readdlm

import PyPlot
using LaTeXStrings
const plt = PyPlot
const mpl = PyPlot.matplotlib

struct Benchmark{Style}
    name     :: String
    filename :: String
    pyplot_style :: Style
    Benchmark(name, fname, style) = new{typeof(style)}(name, fname, style)
end

struct TimerData
    avg :: Vector{Float64}
    std :: Vector{Float64}
    min :: Vector{Float64}
    max :: Vector{Float64}
end

const MPI_TAG = Ref("IntelMPI.2019.9")

const STYLE_IDEAL = (color=:black, ls=:dotted, label="ideal")

function load_timings(bench::Benchmark, resolution)
    filename = joinpath("results", MPI_TAG[], "N$resolution", bench.filename)
    data = readdlm(filename, Float64, comments=true) :: Matrix{Float64}
    Nxyz = data[:, 1:3]
    @assert all(Nxyz .== resolution)
    procs = data[:, 4]
    proc_dims = data[:, 5:6]
    repetitions = data[:, 7]
    times = TimerData((data[:, j] for j = 8:11)...)
    (
        Nxyz = Nxyz,
        procs = procs,
        proc_dims = proc_dims,
        repetitions = repetitions,
        times = times,
    )
end

function plot_from_file!(ax, bench::Benchmark, resolution;
                         plot_ideal=false, error_bars=nothing)
    data = load_timings(bench, resolution)
    times = data.times
    t = times.avg

    ax.plot(data.procs, t; bench.pyplot_style...)

    colour = bench.pyplot_style.color
    if error_bars == :extrema
        ax.errorbar(data.procs, t; color=colour)
    elseif error_bars == :std
        δ = times.std ./ 2
        ax.fill_between(data.procs, t .- δ, t .+ δ; alpha=0.2, color=colour)
    end

    if plot_ideal
        plot_ideal_scaling!(ax, data, t)
        add_text_resolution!(ax, resolution, data.procs, t)
    end

    ax
end

function plot_ideal_scaling!(ax, data, t)
    p = data.procs
    t_ideal = similar(t)
    for n in eachindex(t)
        t_ideal[n] = t[1] * p[1] / p[n]
    end
    ax.plot(p, t_ideal; STYLE_IDEAL...)
end

function add_text_resolution!(ax, N, xs, ys)
    x = first(xs)
    y = first(ys)
    if N == 512
        kws = (ha = :left, va = :top)
        x *= 0.95
        y *= 0.65
    else
        kws = (ha = :right, va = :center)
        x *= 0.9
        y *= 1.02
    end
    ax.text(
        x, y, latexstring("$N^3");
        fontsize = "large",
        kws...
    )
end

function plot_lib_comparison!(ax, benchs, resolution)
    ax.set_xscale(:log, base=2)
    ax.set_yscale(:log, base=10)
    map(benchs) do bench
        plot_ideal = bench === first(benchs)
        plot_from_file!(
            ax, bench, resolution;
            plot_ideal,
            # error_bars = :std,
        )
    end
    ax
end

function legend_libs!(ax, benchs; with_ideal=false, outside=false)
    styles = Any[getfield.(benchs, :pyplot_style)...]
    labels = Any[getfield.(benchs, :name)...]
    if with_ideal
        push!(styles, STYLE_IDEAL)
        push!(labels, "Ideal")
    end
    kws = if outside
        (loc = "center left", bbox_to_anchor = (1.0, 0.5))
    else
        (loc = "lower left", )
    end
    draw_legend!(ax, styles, labels; frameon=false, kws...)
end

function draw_legend!(ax, styles, labels; kwargs...)
    leg = ax.get_legend()
    lines = [mpl.lines.Line2D(Float64[], Float64[]; style...)
             for style in styles]
    ax.legend(lines, labels; kwargs...)
    if leg !== nothing
        ax.add_artist(leg)
    end
    ax
end

# Wrap matplotlib's SVG writer.
struct SVGWriter{File<:IO} <: IO
    fh :: File
end

Base.isreadable(io::SVGWriter) = isreadable(io.fh)
Base.iswritable(io::SVGWriter) = iswritable(io.fh)
Base.isopen(io::SVGWriter) = isopen(io.fh)

function Base.write(io::SVGWriter, s::Union{SubString{String}, String})
    # We remove the image height and width from the header.
    # This way the SVG image takes all available space in browsers.
    p = "\"\\S+pt\""  # "316.8pt"
    pat = Regex("(<svg .*)height=$p (.*) width=$p")
    rep = replace(s, pat => s"\1\2")
    write(io.fh, rep)
end

function plot_timings()
    resolutions = (
        # 256,
        512,
        1024,
        2048,
    )
    style = (fillstyle = :none, ms = 7, markeredgewidth = 1.5, linewidth = 1.5)
    benchs = (
        Benchmark(
            "PencilFFTs (default)", "PencilFFTs_PI.dat",
            (style..., color = :C0, marker = :o, zorder = 5),
        ),
        Benchmark(
            "PencilFFTs (Alltoallv)", "PencilFFTs_PA.dat",
            (style..., color = :C2, marker = :s, zorder = 8,
             linestyle = :dashed, linewidth = 1.0),
        ),
        Benchmark(
            "P3DFFT", "P3DFFT2.dat",
            (style..., color = :C1, marker = :x, zorder = 4, markeredgewidth = 2),
        ),
    )

    fig = plt.figure(figsize = (6, 4.2) .* 1.1, dpi = 150)
    ax = fig.subplots()

    ax.set_xlabel("MPI processes")
    ax.set_ylabel("Time (milliseconds)")

    map(resolutions) do N
        plot_lib_comparison!(ax, benchs, N)
    end

    legend_libs!(ax, benchs, with_ideal=true, outside=false)

    # Show '1024' instead of '2^10'
    for axis in (ax.xaxis, ax.yaxis)
        axis.set_major_formatter(mpl.ticker.ScalarFormatter())
    end

    open("timing_comparison.svg", "w") do ff
        io = SVGWriter(ff)
        @time fig.savefig(io)
    end

    fig
end

plot_timings()
